Applications/FileIO/GocadIO/GocadSGridReader.cpp

```diff
@@ -0,0 +1,855 @@
+    DBUG("x_dim = %d, y_dim = %d, z_dim = %d => #nodes = %d, #cells = %d",
+         x_dim, y_dim, z_dim, _index_calculator._n_nodes,
+         _index_calculator._n_cells);
+}
+
+void GocadSGridReader::parseFileName(std::string const& line,
+                                     std::string& result_string) const
+{
+    boost::char_separator<char> sep(" \r");
+    boost::tokenizer<boost::char_separator<char>> tok(line, sep);
+    auto it(tok.begin());
+    ++it;  // overread POINTS_FILE or FLAGS_FILE or REGION_FLAGS_FILE
+    result_string = _path + *it;
+}
+
+/**
+ * @param line input/output
+ * @param in input stream containing the face set
+ */
+void GocadSGridReader::parseFaceSet(std::string& line, std::istream& in)
+{
+    // create and initialize a Gocad::Property object for storing face set data
+    Gocad::Property face_set_property;
+    face_set_property._property_id = _n_face_sets;
+    face_set_property._property_name = "FaceSet";
+    face_set_property._property_class_name = "FaceSetData";
+    face_set_property._property_unit = "unitless";
+    face_set_property._property_data_type = "double";
+    face_set_property._property_data_fname = "";
+    face_set_property._property_no_data_value = -1.0;
+    face_set_property._property_data.resize(_index_calculator._n_cells);
+    std::fill(face_set_property._property_data.begin(),
+              face_set_property._property_data.end(),
+              face_set_property._property_no_data_value);
+
+    std::istringstream iss(line);
+    std::istream_iterator<std::string> it(iss);
+    // Check first word is FACE_SET
+    if (*it != std::string("FACE_SET"))
+    {
+        ERR("Expected FACE_SET keyword but '%s' found.", it->c_str());
+        throw std::runtime_error(
+            "In GocadSGridReader::parseFaceSet() expected FACE_SET keyword not "
+            "found.");
+    }
+    ++it;
+    face_set_property._property_name += *it;
+    ++it;
+    std::size_t const n_of_face_set_ids(
+        static_cast<std::size_t>(atoi(it->c_str())));
+    std::size_t face_set_id_cnt(0);
+
+    while (std::getline(in, line) && face_set_id_cnt < n_of_face_set_ids)
+    {
+        if (line.back() == '\r')
+        {
+            line.pop_back();
+        }
+        boost::char_separator<char> sep("\t ");
+        boost::tokenizer<boost::char_separator<char>> tokens(line, sep);
+
+        for (auto tok_it = tokens.begin(); tok_it != tokens.end();)
+        {
+            std::size_t id(static_cast<std::size_t>(atoi(tok_it->c_str())));
+            tok_it++;
+            std::size_t face_indicator(
+                static_cast<std::size_t>(atoi(tok_it->c_str())));
+            tok_it++;
+
+            if (id >= _index_calculator._n_nodes)
+            {
+                ERR("Face set id %d is greater than the number of nodes (%d).",
+                    id, _index_calculator._n_nodes);
+            }
+            else
+            {
+                dynamic_cast<MeshLib::GocadNode*>(_nodes[id])
```

Comment:
Either check the result of dynamic cast for nullptr, or (because you can prove it) do a static cast.